### ISSUE TITLE
Data flow considerations

### ISSUE THREAD

**Issue by tlienart** [tlienart/AnalyticalEngine.jl#3]
This package should aim to work on large datasets and with distributed archs. One key element is to be reasonable about how the data is thrown around.

- **bagging, stacking, ...** in those case, several models are fitted on all or part of the data. Potentially this can all happen in parallel. However
  - each model should potentially be considered a "process", if that process happens on a different machine then the data needs to be copied there. Some thought about how to do this in order to not copy around gigabytes of data stupidly would be relevant. For example in bagged model, one may want to just send a bootstrapped **small portion** of the data. 
  - the aggregation of the result should be done seamlessly on the master process, this should be straightforward

- **comparison of pipelines** here for example we may want to consider different pipelines (feature engineering followed by some model fitting), each pipeline would be its own process and each should have access to the (?) entire data potentially requiring to copy things around. 

- **cross validation** same story

So basically here the questions are

- how do we do this well
- how do we make sure the API reflects the possible choices (copying data around, trying to do everything in RAM etc) + advise the user based on data size + RAM size + number of processors (that  would be ideal).

**Comment by fkiraly** [comment]
Suggestion: data provider types and parametric/higher-order learner types

Joint dispatch by data provider and higher-order learner type (which encodes the computation graph) can then take care of data flow under the hood?

**Comment by vollmersj** [comment]
Regarding computation graph aspect. The questions is should we directly work with Dagger @MikeInnes  or will Flux do this in the future.
I have toyed a little bit with it and it should be possible to automatically turn a workflow pipeline into a Dagger graph.

```julia
using Dagger

p = delayed(sin)
q = delayed(cos)(p)
r = delayed(exp)(q)
s = delayed(collect)(p, q, r)
show_plan(s)

using GraphViz
Graph(replace(show_plan(s),"digraph","digraph test"))
```

<img width="119" alt="image" src="https://user-images.githubusercontent.com/12613127/36352270-44dfb646-14ae-11e8-8df4-00f16634356b.png">

**Comment by tlienart** [comment]
@MikeInnes could you maybe clarify what's the difference between [Dagger.jl](https://github.com/JuliaParallel/Dagger.jl) and your [DataFlow.jl](https://github.com/MikeInnes/DataFlow.jl) and which one could be relevant for us?

I guess there are two things

* A. getting some data to a number of processors and training an individual model there then aggregating
* B. training one complicated model across a lot of data (e.g. NN style)

are the two libraries respectively for A / B ?

(*Note*: given that JuliaDB uses Dagger in the backend, I think part of the answer is already there)

**Comment by MikeInnes** [comment]
Probably neither of them are that directly relevant; for (A) you probably just want to use JuliaDB's APIs directly, and all the scheduling will happen under the hood. Dagger is really more aimed at wide-scale transformations of data than scheduling of pipelines for example.

DataFlow.jl is a generic representation for dataflow graphs/programs. It will basically help you if you want to write a compiler / do complex transformations, but it doesn't directly have any kind of scheduling or parallelism. Flux will start to implement gradually more of that stuff though.

**Comment by tlienart** [comment]
Makes sense, thanks Mike particularly the fact that JuliaDB has a scheduler.

I think what we'll need to clarify is how to best interact with it, this is unclear to me at this point bc I have not gone through the julia db codebase in detail. 

I think there's still a bigger picture to be discussed, away from the specific implementation detail, as to how different pieces of models are trained and when, I don't believe JuliaDB by itself is enough for this though I may be wrong.
